refactor: Clean up emoji code for mobile reuse.

showell opened the pull request.

Fingel reviewed (commented):
emoji_utils.js looks good! Just some minor mostly stylistic comments.

timabbott commented on static/js/emoji.js:
```diff
@@ -261,22 +261,37 @@ export function get_emoji_details_for_rendering(opts) {
 
+function build_default_emoji_aliases({names, get_emoji_codepoint}) {
+    // Please keep this as a pure function so that we can
+    // eventually share this code with the mobile codebase.
```
We probably don't need these comments, in that I don't think we're likely to let someone go back to the ugly globals approach in any case :)

showell replied:
All that being said, I don't this to be a huge blocker, so if you just want to snip them or edit them yourself before merging, please do that without another review cycle.

timabbott replied:
I think we can leave them for now; we can drop them when we move the pure functions in a proper shared module down the line.

timabbott commented on static/js/emoji_util.js:
```diff
@@ -0,0 +1,114 @@
+    for (const codepoints of Object.values(emoji_catalog)) {
+        for (const codepoint of codepoints) {
+            const emoji_name = get_emoji_name(codepoint);
+            if (emoji_name !== undefined && !map.has(emoji_name)) {
```
I think if we do that we'd need to reorder, since we're clearly trying to avoid overwriting realm emoji with unicode emoji. 

But I think removing this `has` and then swapping the order would make sense to me.

Fingel replied:
Yup, that's what I had in mind.

showell replied:
Makes sense, this will be in the next push.

showell commented on static/js/emoji_util.js:
```diff
@@ -0,0 +1,114 @@
+    */
+
+    const translations = [];
+    for (const [emoticon, replacement_text] of Object.entries(emoticon_conversions)) {
```
I will leave this cleanup for somebody else, as it feels orthogonal to the main point of the PR, and I would prefer to avoid the churn here.

showell commented:
This is ready for review again.  I addressed the comments above.  Since the last review, I pulled off a commit relating to extracting the pure functions to their own module, and then I added a commit based on the @Fingel suggestion to swap the two loops that build up the emoji map.

timabbott commented:
Merged, thanks @showell!
